include/sunlinsol/sunlinsol_ginkgoblock.hpp

```diff
@@ -0,0 +1,460 @@
+
+      solver_ = solver_factory_->generate(matrix_->GkoMtx());
+    }
+    else { solver_->reset_tolerance(tol); }
+
+    if (!logger_)
+    {
+      logger_ =
+        gko::share(gko::batch::log::BatchConvergence<sunrealtype>::create());
+    }
+    solver_->add_logger(logger_);
+
+    std::unique_ptr<GkoBatchVecType> x_vec{
+      impl::WrapBatchVector(GkoExec(), num_blocks_, x)};
+    std::unique_ptr<GkoBatchVecType> b_vec{
+      impl::WrapBatchVector(GkoExec(), num_blocks_, b)};
+
+    if (scaling_)
+    {
+      // \tilde{b} = S_1 b
+      b_vec->scale(impl::WrapAsMultiVector(col_scale_vec_, num_blocks_));
+    }
+
+    // \tilde{x} = \tilde{A}^{-1} \tilde{b}
+    [[maybe_unused]] gko::batch::BatchLinOp* result =
+      solver_->apply(b_vec.get(), x_vec.get());
+
+    if (scaling_)
+    {
+      // x = S_2^{-1} \tilde{x}
+      x_vec->scale(impl::WrapAsMultiVector(row_scale_vec_, num_blocks_));
+    }
+
+    // Check if any batch entry did not reach the tolerance.
+    bool at_least_one_did_not_converge{false};
+    bool max_res_norm_did_not_reduce{false};
+    sunrealtype max_res_norm{0.0};
+    sunrealtype min_res_norm{SUN_BIG_REAL};
+    gko_exec_->get_master()->copy_from(gko_exec_, num_blocks_,
+                                       logger_->get_residual_norm().get_const_data(),
+                                       res_norm_array_.get_data());
+    auto res_norm = res_norm_array_.get_data();
+
+    for (gko::size_type i = 0; i < num_blocks_; i++)
+    {
+      max_res_norm =
+        std::max(max_res_norm,
+                 res_norm[i]); // NOLINT(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+      min_res_norm =
+        std::min(min_res_norm,
+                 res_norm[i]); // NOLINT(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+    }
+    if (max_res_norm > tol) { at_least_one_did_not_converge = true; }
+    if (max_res_norm >= previous_max_res_norm_)
+    {
+      max_res_norm_did_not_reduce = true;
+    }
+
+    // Compute the average number of iterations across all batch entries
+    // as well as the maximum and minimum iteration counts.
+    gko_exec_->get_master()->copy_from(gko_exec_, num_blocks_,
+                                       logger_->get_num_iterations().get_const_data(),
+                                       iter_count_array_.get_data());
+    auto iter_count = iter_count_array_.get_data();
+
+    sunindextype max_iter_count{0};
+    sunindextype min_iter_count{max_iters_};
+    avg_iter_count_ = 0.0;
+    for (gko::size_type i = 0; i < num_blocks_; i++)
+    {
+      avg_iter_count_ +=
+        iter_count[i]; // NOLINT(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+      max_iter_count =
+        std::max(max_iter_count,
+                 iter_count[i]); // NOLINT(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+      min_iter_count =
+        std::min(min_iter_count,
+                 iter_count[i]); // NOLINT(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+    }
```

Comment:
Can consider `std::accumulate` and `std::minmax_element` here

Reply:
Should all of these statistics be computed each solve? Can they be moved into `AvgNumIters` and related functions so users only pay the cost of computing the stats if they request them (or have logging on)?

Reply:
I profiled this, and the cost of computing these every time was negligible cost.

Reply:
I will note this does require a device to host transfer since the Ginkgo logger stores the stats in device memory